Implement InternalObserver for Observable

keithamus opened the pull request:
#### 14125a03f1ad98b91fbe1d65d46c6f15c36e421b
<pre>
Implement InternalObserver for Observable
<a href="https://bugs.webkit.org/show_bug.cgi?id=276839">https://bugs.webkit.org/show_bug.cgi?id=276839</a>

Reviewed by Ryan Reno.

This implements `InternalObserver` for the `Observable` and
`Subscriber` classes, simplifying their functionality by
using a delegate class to manage some of the different types
of Observers. This will make it much easier to implement
operators (e.g. take, map, filter).

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
(WebCore::JSSubscriber::visitAdditionalChildren):
* Source/WebCore/dom/InternalObserver.cpp: Copied from Source/WebCore/dom/Observable.h.
(WebCore::InternalObserver::error):
* Source/WebCore/dom/InternalObserver.h: Copied from Source/WebCore/dom/Observable.h.
(WebCore::InternalObserver::complete):
(WebCore::InternalObserver::InternalObserver):
  The point of this class is to act as a base for further operators which will
  require mapping observables by creating intermediary subsribers, each which
  need an Observer. The spec describes these as Internal Observer.
  (<a href="https://wicg.github.io/observable/#internal-observer)">https://wicg.github.io/observable/#internal-observer)</a>
* Source/WebCore/dom/InternalObserverFromScript.cpp: Added.
(WebCore::InternalObserverFromScript::create):
(WebCore::InternalObserverFromScript::next):
(WebCore::InternalObserverFromScript::error):
(WebCore::InternalObserverFromScript::complete):
(WebCore::InternalObserverFromScript::visitAdditionalChildren const):
(WebCore::InternalObserverFromScript::InternalObserverFromScript):
  This class allows us to simpify Subscriber so that it only ever takes
  InternalObservers, by taking the existing way Subscriber consumes a
  SubscriptionObserver and mapping that to an InternalObserver delegate instead.
  I called this InternalObserverFromScript as it&apos;s InternalObserver that will
  only be created from the Observer#subscribe() IDL calls.
* Source/WebCore/dom/InternalObserverFromScript.h: Copied from Source/WebCore/dom/Observable.h.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::subscribe):
(WebCore::Observable::subscribeInternal):
(WebCore::Observable::Observable):
(WebCore::Observable::makeSubscriber): Deleted.
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::create):
(WebCore::Subscriber::Subscriber):
(WebCore::Subscriber::next):
(WebCore::Subscriber::error):
(WebCore::Subscriber::complete):
* Source/WebCore/dom/Subscriber.h:

Canonical link: <a href="https://commits.webkit.org/281278@main">https://commits.webkit.org/281278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/224499025d374806bfc09e81df3ffab528d2e02d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59323 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63237 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9766 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61452 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48175 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6948 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61353 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36123 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51344 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28998 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32834 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8600 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8770 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64970 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8808 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55526 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3262 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51342 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55617 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13151 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2705 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34482 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35565 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36651 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35310 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

keithamus commented on Source/WebCore/dom/InternalObserver.h:
```diff
@@ -0,0 +1,131 @@
+
+namespace WebCore {
+
+class InternalObserver : public ActiveDOMObject, public RefCounted<InternalObserver> {
```
The point of this class is to act as a base for further operators which will require mapping observables by creating intermediary subsribers, each which need an Observer. The [spec describes these as Internal Observer](https://wicg.github.io/observable/#internal-observer).

keithamus commented on Source/WebCore/dom/InternalObserver.h:
```diff
@@ -0,0 +1,131 @@
+};
+
+
+class InternalObserverFromScript final : public InternalObserver {
```
This class allows us to simpify `Subscriber` so that it only ever takes `InternalObserver`s, by taking the existing way `Subscriber` consumes a  `SubscriptionObserver` and mapping that to an `InternalObserver` delegate instead. I called this `InternalObserverFromScript` as it's `InternalObserver` that will only be created from the `Observer#subscribe()` IDL.

rreno replied:
These comments you've  made would be great to add to the commit message. I like to use the changed files section for commentary like this.

e.g..

```
* Source/WebCore/dom/InternalObserver.h: Added
(WebCore::InternalObserver::complete):
(WebCore::InternalObserver::error):
(WebCore::InternalObserver::InternalObserver):
    <my comments here>
```

keithamus commented on Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
```diff
@@ -31,12 +31,6 @@ namespace WebCore {
 template<typename Visitor>
 void JSSubscriber::visitAdditionalChildren(Visitor& visitor)
 {
-    if (auto* next = wrapped().nextCallbackConcurrently())
-        next->visitJSFunction(visitor);
-    if (auto* error = wrapped().errorCallbackConcurrently())
-        error->visitJSFunction(visitor);
-    if (auto* complete = wrapped().completeCallbackConcurrently())
-        complete->visitJSFunction(visitor);
```
`Subscriber` no longer holds onto these, instead `InternalObserverFromScript` does. Is there a way I should traverse and mark the objects held in that class?

rreno replied:
You would probably need to call something like `wrapped().observer().visitAdditionalChildren` and then you can define the visitation function on the `InternalObserver` object, keeping in mind thread safety.

rreno replied:
There's a few ways to do it so you'd want to consider which objects are keeping other objects alive and then use the owning object to manage the lifetime of the owned object.

keithamus replied:
One concern I had with this is that it's not possible to build virtual template functions, so I was unsure how this could be done with each specialisation.

rreno replied:
You'd need to do what `ActiveDOMCallback` does and define them as overloads instead of templates.

rreno commented on Source/WebCore/dom/InternalObserver.h:
```diff
@@ -0,0 +1,132 @@
+    void error(JSC::JSValue error) final
+    {
+        if (m_error)
+            m_error->handleEvent(error);
```
```suggestion
           return m_error->handleEvent(error);
       InternalObserver::error(error);
```

rreno commented on Source/WebCore/dom/InternalObserver.h:
```diff
@@ -0,0 +1,132 @@
+    void deref() const final { RefCounted::deref(); }
+
+protected:
+    bool m_active = true;
```
```suggestion
    bool m_active { true };
```

rreno commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -25,33 +25,53 @@
+        WTF::switchOn(
+            observer.value(),
+            [&](RefPtr<JSSubscriptionObserverCallback>& next) {
+                subscribeInternal(context, adoptRef(*new InternalObserverFromScript(context, next)), options);
```
These aren't really FromScript, since they're being created here in C++. Does it make sense to have InternalObserver be pure virtual if there's only one class which inherits from it?

keithamus replied:
It looks a bit silly in this PR but subsequent to this, each operator (e.g. take, takeUntil, filter, map) will each implement their own InternalObserver specialisation.

rreno commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -25,33 +25,53 @@
+        WTF::switchOn(
+            observer.value(),
+            [&](RefPtr<JSSubscriptionObserverCallback>& next) {
+                subscribeInternal(context, adoptRef(*new InternalObserverFromScript(context, next)), options);
```
Also, please make a static `InternalObserver::create` function which does the `adoptRef` call (and I would also suggest putting the `suspendIfNeeded` call there instead of in the constructor).

e.g.

```C++
Ref<InternalObserver> InternalObserver::create(<args>)
{
    Ref internalObserver = adoptRef(*new InternalObserver(<args>));
    internalObserver->suspendIfNeeded();
    return internalObserver;
}
```

rreno commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -71,28 +91,12 @@ void Observable::subscribe(ScriptExecutionContext& context, std::optional<Observ
+Observable::Observable(RefPtr<SubscriberCallback> subscriber)
+    : m_subscriber(subscriber)
 {
+    ASSERT(m_subscriber);
```
Can an observable be without a subscriber callback? If not, this should be a `Ref<Subscriber>`

keithamus replied:
For now it cannot, follow up PRs which add new operators will pass in a lambda instead of the `Subscriber` object.

rreno replied:
Still, if the invariant is that this cannot be null then instead of a RefPtr it should be a Ref.

rreno commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -25,33 +25,53 @@
     if (options.signal)
         subscriber->followSignal(*options.signal.get());
 
+    ASSERT(m_subscriber);
```
`m_subscriber` is the callback, right? I think given we have a `subscriber` and a `m_subscriber` which are two different types the name of this variable can be improved. I suggest `m_subscriberCallback`.

rreno commented on Source/WebCore/dom/InternalObserver.h:
```diff
@@ -0,0 +1,164 @@
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        m_next->visitJSFunction(visitor);
```
These need null checks in both overloads.

rreno commented on Source/WebCore/dom/InternalObserver.h:
```diff
@@ -0,0 +1,164 @@
+#include "SubscriptionObserver.h"
+#include <wtf/RefCounted.h>
+
+namespace WebCore {
```
As an overall comment on this file I think its best if we create a cpp file to put the implementation into. I'm not sure we gain anything having everything defined inline.

rreno commented on Source/WebCore/dom/Observable.cpp:
```diff
@@ -25,29 +25,47 @@
 
 void Observable::subscribe(ScriptExecutionContext& context, std::optional<ObserverUnion> observer, SubscribeOptions options)
+{
+    if (observer.has_value()) {
```
nit: We could just use `if (observer)` since `optional` has an `operator bool`.

rreno reviewed (approved):
r=me assuming bots are green. Thanks!

keithamus committed:
Implement InternalObserver for Observable
https://bugs.webkit.org/show_bug.cgi?id=276839

Reviewed by Ryan Reno.

This implements `InternalObserver` for the `Observable` and
`Subscriber` classes, simplifying their functionality by
using a delegate class to manage some of the different types
of Observers. This will make it much easier to implement
operators (e.g. take, map, filter).

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSSubscriberCustom.cpp:
(WebCore::JSSubscriber::visitAdditionalChildren):
* Source/WebCore/dom/InternalObserver.cpp: Copied from Source/WebCore/dom/Observable.h.
(WebCore::InternalObserver::error):
* Source/WebCore/dom/InternalObserver.h: Copied from Source/WebCore/dom/Observable.h.
(WebCore::InternalObserver::complete):
(WebCore::InternalObserver::InternalObserver):
  The point of this class is to act as a base for further operators which will
  require mapping observables by creating intermediary subsribers, each which
  need an Observer. The spec describes these as Internal Observer.
  (https://wicg.github.io/observable/#internal-observer)
* Source/WebCore/dom/InternalObserverFromScript.cpp: Added.
(WebCore::InternalObserverFromScript::create):
(WebCore::InternalObserverFromScript::next):
(WebCore::InternalObserverFromScript::error):
(WebCore::InternalObserverFromScript::complete):
(WebCore::InternalObserverFromScript::visitAdditionalChildren const):
(WebCore::InternalObserverFromScript::InternalObserverFromScript):
  This class allows us to simpify Subscriber so that it only ever takes
  InternalObservers, by taking the existing way Subscriber consumes a
  SubscriptionObserver and mapping that to an InternalObserver delegate instead.
  I called this InternalObserverFromScript as it's InternalObserver that will
  only be created from the Observer#subscribe() IDL calls.
* Source/WebCore/dom/InternalObserverFromScript.h: Copied from Source/WebCore/dom/Observable.h.
* Source/WebCore/dom/Observable.cpp:
(WebCore::Observable::subscribe):
(WebCore::Observable::subscribeInternal):
(WebCore::Observable::Observable):
(WebCore::Observable::makeSubscriber): Deleted.
* Source/WebCore/dom/Observable.h:
* Source/WebCore/dom/Subscriber.cpp:
(WebCore::Subscriber::create):
(WebCore::Subscriber::Subscriber):
(WebCore::Subscriber::next):
(WebCore::Subscriber::error):
(WebCore::Subscriber::complete):
* Source/WebCore/dom/Subscriber.h:

Canonical link: https://commits.webkit.org/281278@main
